### DISCUSSION TITLE
Failed to compile tsh on Android

### DISCUSSION THREAD

**Discussion by adoyle-h** [gravitational/teleport#31481] [Q&A] [answered]
I want to use `tsh` on Android device with Termux. The Linux ARM64/ARMv8 (64-bit) version not work. So I build it from source, `git clone` and `make build/tsh`. But it shows 1 error.

```sh
> make $BUILDDIR/tsh
GOOS=android GOARCH=arm64 CGO_ENABLED=1 go build -tags "   " -o /data/data/com.termux/files/usr/bin//tsh  -ldflags '-w -s' -trimpath ./tool/tsh
# github.com/gravitational/teleport/lib/inventory/metadata
lib/inventory/metadata/metadata_linux.go:22:11: fatal error: 'gnu/libc-version.h' file not found
 #include <gnu/libc-version.h>
          ^~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [Makefile:308: /data/data/com.termux/files/usr/bin//tsh] Error 1
```

I also tried to cross-compile on MacOS with Android NDK. Get the same error.

```sh
> CC=/Applications/AndroidNDK9519653.app/Contents/NDK/toolchains/llvm/prebuilt/darwin-x86_64/bin/aarch64-linux-and
roid33-clang GOOS=android GOARCH=arm64 CGO_ENABLED=1 go build -tags "   " -o build/tsh  -ldflags '-w -s' -trimpath ./tool/tsh
# github.com/gravitational/teleport/lib/inventory/metadata
lib/inventory/metadata/metadata_linux.go:22:11: fatal error: 'gnu/libc-version.h' file not found
 #include <gnu/libc-version.h>
          ^~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by webvictim** [comment]
Teleport requires `glibc` to compile and run. Android doesn't use `glibc`, it uses [Bionic](https://en.wikipedia.org/wiki/Bionic_(software)). As such, Teleport can't currently be compiled for Android.

**Reply by adoyle-h** [reply]
Is there any plan to implement it?

**Comment by Shareed2k** [comment, marked as answer]
you can just apply patch for this file `lib/inventory/metadata/metadata_linux.go`, i'm already prepared [gist](https://gist.github.com/Shareed2k/f0bdab37f8cb93e320a4e09990b9bb71) with the patch, this metadata is unnecessary for use the tool
```sh
curl -s https://gist.githubusercontent.com/Shareed2k/f0bdab37f8cb93e320a4e09990b9bb71/raw/93b7f434d952f09d9cb70c43c2b5395dbc384310/metadata_linux.go.patch | git apply
```

**Reply by adoyle-h** [reply]
It works! Thank you so much.

**Reply by Shareed2k** [reply]
https://github.com/termux/termux-packages/pull/17852
